sqlmesh/core/engine_adapter/base.py

```diff
@@ -144,6 +145,51 @@ def cursor(self) -> t.Any:
     def spark(self) -> t.Optional[PySparkSession]:
         return None
 
+    @classmethod
+    def is_pandas_df(cls, value: t.Any) -> bool:
+        return isinstance(value, pd.DataFrame)
+
+    def _to_source_queries(
+        self,
+        query_or_df: QueryOrDF,
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]],
+        *,
+        target_table: t.Optional[TableName] = None,
+        batch_size: t.Optional[int] = None,
+    ) -> t.Tuple[t.List[SourceQuery], t.Optional[t.Dict[str, exp.DataType]]]:
+        batch_size = self.DEFAULT_BATCH_SIZE if batch_size is None else batch_size
+        if isinstance(query_or_df, (exp.Subqueryable, exp.DerivedTable)):
+            return [SourceQuery(query_factory=lambda: query_or_df)], columns_to_types  # type: ignore
+        else:
+            return self._df_to_source_queries(
+                query_or_df, columns_to_types, batch_size, target_table=target_table
+            )
+
+    def _df_to_source_queries(
+        self,
+        df: DF,
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]],
+        batch_size: int,
+        target_table: t.Optional[TableName] = None,
+    ) -> t.Tuple[t.List[SourceQuery], t.Optional[t.Dict[str, exp.DataType]]]:
+        assert isinstance(df, pd.DataFrame)
+        num_rows = len(df.index)
+        columns_to_types = columns_to_types or columns_to_types_from_df(df)
```

Comment:
what's going on here? why do you have to return columns_to_types? why is that coupled in this function

Reply:
In a follow up PR `_to_source_queries` will be returning a non-optional columns_to_types. So this is building towards that.

To answer why is it coupled, when we get dataframes we can infer the columns_to_types (if not provided) based on the schema of that Dataframe. The only exception right now is PySpark. Since this is where we have the logic to convert a dataframe to SQL, and it often requires knowing columns to type, we should have it also return the columns_to_types that it inferred.